### ISSUE TITLE
Minimal biolinkml version attribute

### ISSUE THREAD

**Issue by hsolbrig** [biolink/biolinkml#308]
BiolinkML schemas need the ability to state the minimum biolinkml version required to interpret the model.  It would be a lot nicer to say "Sorry: I'm not smart enough to understand what's in this model" instead of regurgitating pages of error messages on perfectly valid models.

**Comment by RichardBruskiewich** [comment]
@hsolbrig,  @deepakunni3 and I are thinking of adding  a `metamodel_version:`  tag  to the biolink-model.yaml  file to  explicitly declare the version of  BiolinkML required to  understand the model.  We'd change the BiolinkML parsing to suit.

**Comment by hsolbrig** [comment]
Ya - I think that this definitely needs to be part of the core.   Would propose a label more along the line of `requires_version:` or something similar.  Might be useful to use the same semantics as the pip requirements?

**Comment by RichardBruskiewich** [comment]
@hsolbrig (cc: @deepakunni3,  @cmungall),  would `requires_version:`  be too ambiguous in the context of the Biolink Model  which  already  had  a  simple `version:` tag?

Distinguishing between  the two contexts (main and  'meta') is  why  `metamodel_version` seems a bit more obvious?

**Comment by hsolbrig** [comment]
`version` is an optional property, as intended to represent the version of a *model*.   Examples:

* https://github.com/biolink/biolink-model/blob/master/biolink-model.yaml#L7 -- the current version of the biolink model is 1.4.1
* https://github.com/biolink/biolinkml/blob/master/meta.yaml#L6 -- the current version of the biolink metamodel is 1.6.1

`metamodel_version` has been around since the beginning.  It records the version of the metamodel that was used to process a model.  Examples:

* https://github.com/biolink/biolink-model/blob/master/biolink/model.py#L28 - metamodel version 1.6.1 was used to generate the python representation of biolink-model
* https://github.com/biolink/biolinkml/blob/master/biolinkml/meta.py#L29 -- metamodel version 1.6.1 was used to generate the python representation of biolinkml

The bit that is missing, however, is some sort of declaration in the model that says what the minimum metamodel version required to understand a given metamodel must be.  As an example, version 1.7.0 of biolinkml, when released, will support a new collection called `enums`.   If `biolink-model` takes advantage of this, any metamodel earlier than `1.7.0` won't be able to process an `enums` collection.  `"requires_version": ">=1.7.0"` could record this fact.

**Comment by hsolbrig** [comment]
I might add, however, that we do have an issue in the queue (note to self: find this) to clean up some of these versions.  We've got too many floating around now, including:

* metamodel version  (discussed above)
* model version (discussed above)
* specific generator version (yamlgen, for example, is "0.1.0" https://github.com/biolink/biolinkml/blob/master/biolinkml/generators/yamlgen.py#L17 and pythongen is at "0.9.0" https://github.com/biolink/biolinkml/blob/master/biolinkml/generators/__init__.py
* github/pypi release version -- this is set by the release tag (https://github.com/biolink/biolinkml/releases).

The first two versions (`metamodel_version` and `version` are actively used).  The specific generator version number has never been maintained and we may want to consider removing that attribute completely -- the notion that the generators and metamodel come as a complete package.  The last version is _manually_ synchronized with the metamodel version.  It would be useful to, at bare minimum, add some sort of test that makes sure that that synchronization remails

**Comment by RichardBruskiewich** [comment]
Season's Greetings @hsolbrig,

Deepak did assign me the task of somehow start tracking the BiolinkML version somewhere in the Biolink Model, but I must confess to feeling somewhat confused by all the distinct versioning facets above. If I'm confused, I'm sure that the users of the (Biolink) model may be even worse off. 

Perhaps @deepakunni3 and @cmungall can chime in on this issue and decide exactly how to record this model metadata.  It would be very nice to avoid being too generic on how we tag things here... `version:` and `required_version:` don't now seem very precise.

Given the many tentacled nature of the beast, I wonder if a dedicated structured YAML field for all versioning and version dependencies may be useful here? For example:

```
   version:
       model: 1.4.1
       metamodel: 1.7.0
       pypi:  1.7.0
       pythongen: 0.1.0
```

P.S. not sure about the divergence between BiolinkML _metamodel_ and _pypi_. Are there circumstances where they would not be the same?

**Comment by cmungall** [comment]
the structured field feels a bit too much, and I'm not sure of the use case
for recording more than model and metamodel version. I think the specific
generator versions are too much, I don't know why we are tracking these, is
it up to individual developers to update these? I am not sure I have
updated this e.g when I update owlgen

**Comment by RichardBruskiewich** [comment]
OK @hsolbrig and @cmungall, 

Just the model and metamodel version then. We have the `version` tag for the model. I'm unconfortable with simply calling the metamodel tag `required version`. Would `required metamodel` be less ambigous?

**Comment by hsolbrig** [comment]
Just to be clear -- we toss the generator version (which I agree with).  We then have three attributes:
1) `version` -- the version of the model (e.g. biolink-model, ssom model, etc).  Supplied by model author -- optional or required?
2) `metamodel_version` -- the version of of LinkML that was used to generate an artifact.  Readonly -- supplied by LinkML in the generation phase.
3) `requires_metamodel_version`  -- the minimum metamodel_version needed to correctly process the model.  Supplied by the author, optional.  Do we get clever w/ the syntax and use the requirements.txt syntax? 

Even as I ask question 3), I find myself wondering whether we shouldn't just use requirements.txt/pipenv to control this.  If, for instance, the biolink_model requires > 1.7.0, wouldn't it make just as much sense to say that in the setup?  Otherwise, what would `"requires_version": "1.6.10"` mean if, in the requirements file, you've got `biolinkml="~1.7"`?

**Comment by cmungall** [comment]
I think you are saying we only need attributes for 1+2

3 is already covered, e.g.
https://github.com/biolink/biolink-model/blob/master/requirements.txt#L13

**Comment by deepakunni3** [comment]
Yes, I agree with Chris that we already control the required metamodel version via requirements.txt

So we can make do with `version` and `metamodel_version`.

**Comment by RichardBruskiewich** [comment]
> BiolinkML schemas need the ability to state the minimum biolinkml version required to interpret the model. It would be a lot nicer to say "Sorry: I'm not smart enough to understand what's in this model" instead of regurgitating pages of error messages on perfectly valid models.

It's still the 'cart before the horse' issue, but perhaps, a moot one? 

What exactly do we mean by _"biolinkml version required to interpret the model"_?

Generally speaking, as we've mentioned, the pip `requirements.txt ` file stipulates the version processing constraints... that value is manually specificed by the creator of the model project. Is not the action of running a specified version of BiolinkML for parsing is explicitly (hence knowingly) done by the creator of the schema? 

Is there a point *after* BiolinkML is run where end users of the results _need to know_ which BiolinkML metamodel version was used, in order to constrain model interpretation or rather, does the use of the specific model artifacts generated by the run of the (requirements.txt) specified BiolinkML version drive the interpretation?

I guess it is painless to simply document the `metamodel_version` used by the run, in the generated artifacts (not even, the schema YAML)?

**Comment by hsolbrig** [comment]
> What exactly do we mean by "biolinkml version required to interpret the model"?
The hint that allows the software to say "Sorry: I'm not smart enough..."

Good points -- without something IN the model, it would be hard to understand why regurgitation was occurring, so, indeed, it might be handy to have something in the model itself after all which, in turn, leads back to the question of what do we call it.  `requires_metamodel_version`?  `minimum_metamodel_version`?  `depends_on_metamodel_version`?

**Comment by hsolbrig** [comment]
"Thus, it is those artifacts which can contain the metamodel version tag (e.g. in a standard but parseable artifact-format complaint comment or toplevel variable?) which can be inspected by the system using the artifact."

This already exists - it is available to any generator and many already emit it in the output.  Example: https://github.com/biolink/biolinkml/blob/552b5f42f7803e765f7cf3e35f4aa73224b0fd84/tests/test_issues/output/issue_107.py#L29

Am I missing something?

**Comment by RichardBruskiewich** [comment]
Nope.

Just that perhaps our final decision is that perhaps, we've come full circle with your original issue statement: no action required given that:

1. Projects with schemata being processed by BiolinkML must have some explicit statement (i.e. in pip requirements.txt) of the minimal (or better) BiolinkML model required to run to generate the schema artifacts that schema users will exploit in their systems using the schema.

2. That the current BiolinkML practice of tagging artifacts with the BiolinkML (metamodel) version used suffices for the practical consequences of this issue concern. No need to mark up the YAML file itself (not useful nor necessary - requirements.txt covers us there).

The endpoint of this issue is simply that we now better understand our BiolinkML (metamodel) version tracking needs?  Suggests  that we simply better document this understanding somewhere centrally (if needed) then close this issue: no further action required?
